File: public/dsa/two_pointers/maximum_k_substring_vowels/optimal/maximum_k_substring_vowels.cpp
#include <bits/stdc++.h>
using namespace std;

bool is_vowel(char c) {
    return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u';
}

int maxKSubstringVowels(string s, int k) {
    int max_vowels = 0;
    int vowels = 0;
    int n = s.size();

    for (int i = 0; i < n; i++) {
        if (i >= k && is_vowel(s[i - k])) {
            vowels--;
        }
        if (is_vowel(s[i])) {
            vowels++;
        }
        if (i >= k - 1) {
            max_vowels = max(max_vowels, vowels);
        }
    }

    return max_vowels;
}

// tests-start
int main() {
    string s1 = "workaattech";
    int k1 = 3;
    cout << maxKSubstringVowels(s1, k1) << endl;

    string s2 = "substring";
    int k2 = 2;
    cout << maxKSubstringVowels(s2, k2) << endl;

    return 0;
}
// tests-end

/*output
2
1
*/

File: public/dsa/two_pointers/maximum_k_substring_vowels/optimal/maximum_k_substring_vowels.py


def is_vowel(c):
    return c in ['a', 'e', 'i', 'o', 'u']


class Solution:
    def maxKSubstringVowels(self, s: str, k: int) -> int:
        max_vowels = 0
        vowels = 0
        for i in range(len(s)):
            if i >= k:
                if is_vowel(s[i-k]):
                    vowels -= 1
            if is_vowel(s[i]):
                vowels += 1
            if i >= k - 1:
                max_vowels = max(max_vowels, vowels)
        return max_vowels


# tests-start
sol = Solution()
s = "workaattech"
k = 3
print(sol.maxKSubstringVowels(s, k))
s = "substring"
k = 2
print(sol.maxKSubstringVowels(s, k))
# tests-end

'''output
2
1
'''
